File: app/src/main/java/me/eagzzycsl/darkroom/ui/BaseAdapter.java
package me.eagzzycsl.darkroom.ui;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.util.ArrayList;

import me.eagzzycsl.darkroom.model.MyApp;

abstract class BaseAdapter<M extends MyApp> extends RecyclerView.Adapter<BaseAdapter<M>.RecViewHolder<M>> {
    protected Context context;
    private ArrayList<M> dataArray;

    BaseAdapter(Context context) {
        this.context = context;
    }

    @Override
    public void onBindViewHolder(RecViewHolder<M> holder, int position) {
        holder.setContent(dataArray.get(position));
    }

    @Override
    public int getItemCount() {
        return dataArray == null ? 0 : dataArray.size();
    }

    public abstract int getLayoutId();

    void setDataArray(ArrayList<M> dataArray) {
        this.dataArray = dataArray;
    }

    abstract class RecViewHolder<E extends MyApp> extends RecyclerView.ViewHolder {
        RecViewHolder(View itemView) {
            super(itemView);
        }

        public abstract void setContent(E data);
    }

}

File: app/src/main/java/me/eagzzycsl/darkroom/ui/BaseAdapter.kt
package me.eagzzycsl.darkroom.ui

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.View

import java.util.ArrayList

abstract class BaseAdapter<MetaApp>(var context: Context)
    : RecyclerView.Adapter<BaseAdapter<MetaApp>.RecViewHolder<MetaApp>>() {
    private var dataArray: ArrayList<MetaApp> = ArrayList()

    abstract val layoutId: Int

    override fun onBindViewHolder(holder: RecViewHolder<MetaApp>, position: Int) {
        holder.setContent(dataArray[position])
    }

    override fun getItemCount(): Int {
        return dataArray.size
    }

    fun setDataArray(dataArray: ArrayList<MetaApp>) {
        this.dataArray = dataArray
    }

    abstract inner class RecViewHolder<in MetaApp>(itemView: View) : RecyclerView.ViewHolder(itemView) {

        abstract fun setContent(data: MetaApp)
    }

}